hanish/bot.py

```diff
@@ -0,0 +1,409 @@
+
+        # If we get this far without a botid, we have a problem.
+        raise SlackException(
+            "could not find an ID for a bot named {}".format(self.name)
+        )
+
+    @memoized
+    def atbotid(self):
+        """
+        Computes and memoizes the @botid regular expression for directed
+        messages that may contain commands. E.g. the pattern for <@botid>.
+        """
+        return re.compile(
+            re.escape("<@{}>".format(self.botid)), re.I
+        )
+
+    def run(self):
+        """
+        Connects to the Slack real-time messaging API, polling every second
+        for messages
+        """
+        # Set up the signal handlers
+        # TODO: use threads to better manage communication
+        self.shutdown = False
+        signal.signal(signal.SIGINT, lambda signal, frame: self.stop())
+
+        # Fetch and cache the current weather at startup
+        self.weather()
+
+        # Schedule timed events
+        self.set_schedule()
+
+        # Connect to the real time message API
+        if self.slack.rtm_connect():
+            # TODO: add better chat logging
+            print("slackbot named {} ({}) is connected".format(self.name, self.botid))
+
+            while True:
+
+                # Check if we're shutdown
+                if self.shutdown:
+                    break
+
+                # Check to see if anything has come through the channel
+                self.read_rtm_channel()
+
+                # Run any scheduled tasks
+                schedule.run_pending()
+
+                # Sleep for a bit until the next time we read the channel
+                time.sleep(POLLING_INTERVAL)
+
+        else:
+            raise SlackException(
+                "could not connect to API: invalid Slack API key or Bot ID?"
+            )
+
+        # If we've made it here, we've been shutdown
+        # TODO: convert to logging functionality
+        print("bot has been gracefully shutdown")
+
+    def stop(self):
+        """
+        Set the shutdown semaphore
+        """
+        self.shutdown = True
+
+    def read_rtm_channel(self):
+        """
+        Reads the RTM channel, parses output, filters messages based on type
+        and if they are directed at the bot. If so, passes those messages on
+        to the message handler command.
+        """
+
+        # Check for messages on the wire
+        # TODO: as soon as we connect, the last message comes through.
+        messages = self.slack.rtm_read()
+        if messages and len(messages) > 0:
+            # If there are messages, begin filtering and parsing.
+            for msg in messages:
+                # Filter only messages from other users
+                if msg["type"] == "message":
+                    # Check if an @botid appears in the message
+                    if self.atbotid.search(msg.get("text", "")):
+                        # Handle the atbotid message
+                        self.handle_message(msg)
```

Comment:
So the comprehension in this case would create a new list in memory containing the output of `handle_message` (in this case None) whose length is the number of messages. Generators, filters, and `yield` is all about reducing the amount of memory being used while iterating over the collection. Per your suggestion, you might see something like this in my code elsewhere:

```python 
# Apply filters 
messages = filter(lambda m: m["type"] == "message", messages)
messages = filter(lambda m: self.atbotid.search(m.get("text", "")), messages)
# add more filters as needed 

# Execute message handling 
for message in messages: 
     self.handle_message(message)
```

Which also gives us the possibility of parallelizing message handling (though this would be problematic in a chatbot case). Or we could apply arbitrary filters:

```python
for ffunc in self.filters:
    messages = filter(ffunc, messages) 
```

The `filter` function returns a generator, which is an unexecuted iterable. It is only evaluated in the final for loop. Note that `map(self.handle_message, messages)` won't execute because `map` also returns a generator and needs to be iterated on in order to be executed. Since we're not collecting output, the normal way of doing this, `list(map(self.handle_message, messages))` doesn't make sense. 

Honestly, I didn't do it this way because I was trying to show more readable code and didn't want you to think I was trying to trick you or be overly clever.